### PR TITLE
Migrate to conan ci

### PR DESCRIPTION
**Pull request by steakhal**
## This PR does the following things:
 - uses conan to acquire Catch2 if and only if the `nameof` lib is build using the `build_test` option
 - uses Appveyor and travis to make sure that the `nameof` library is usable on certain configuration and all the tests pass and the example builds
 - it will upload the package (recipe, nameof.hpp and the generated nameofConfig,cmake ... files) to bintray with the version specified in the `conanfile.py`.

## Note that
The CI services can run any time you push a commit, it won't upload any versions to bintray, only if you create a new tag (with any name pattern) like `v0.9.1`
It won't upload any built binaries (neither tests, nor example), but it will upload the sources of the tests and example, since a consumer may want to build it himself.

### CODE REVIEW

**Comment by Neargye** [comment]
Hi, sorry I can check it only tomorrow.

**Comment by steakhal** [comment]
No problem, its open source. :D Nobody paid for doing this, take your time.

**Comment by Neargye** [comment]
Could you please return the original name appveyor?  This is a purely aesthetic thing, I like it when ci files are hidden in dir.

**Comment by steakhal** [comment]
Sounds reasonable. I will rename appveyor.yml to `.appveyor.yml`.

File: .travis.yml [hunk 1]
```diff
@@ -1,162 +1,63 @@
...
-  - cmake -G "Unix Makefiles" -DCMAKE_PREFIX_PATH=../Catch2/installed ..
+install:
+  - chmod +x .ci/install.sh
+  - ./.ci/install.sh
 
 script:
-  - cmake --build . -- -j${JOBS}
-  - ctest --output-on-failure -j${JOBS}
-
-notifications:
```
Comment (Neargye) [inline .travis.yml]: Need return
`notifications:
  email: false`
To disable email notif.

File: .travis.yml [hunk 2]
```diff
@@ -1,162 +1,63 @@
...
-      travis_retry wget -O - ${LIBCXXABI_URL} | tar --strip-components=1 -xJ -C llvm/projects/libcxxabi
-      (cd llvm/build && cmake .. -DCMAKE_INSTALL_PREFIX=${LLVM_INSTALL})
-      (cd llvm/build/projects/libcxx && sudo make install -j${JOBS})
-      (cd llvm/build/projects/libcxxabi && sudo make install -j${JOBS})
-      export CXXFLAGS="-isystem ${LLVM_INSTALL}/include/c++/v1"
-      export LDFLAGS="-L ${LLVM_INSTALL}/lib -l c++ -l c++abi"
-      export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${LLVM_INSTALL}/lib"
-    fi
+      - <<: *osx
+        osx_image: xcode10.1
```
Comment (Neargye) [inline .travis.yml]: Maybe add Xcode 1.2 or 1.3? Something more actual.

File: .travis.yml (hunk 1 above)
Comment (steakhal) [reply .travis.yml]: That's a very good idea :D

Comment (Neargye) [reply .travis.yml]: I found how disable mail in settings on site.

File: test_package/test_package.cpp [hunk 3]
```diff
@@ -14,15 +14,14 @@ int main() {
   static_assert("structvar" == name);
 
   // Nameof.
-  using std::literals::string_view_literals::operator""sv;
+  using std::literals::string_view_literals;
```
Comment (steakhal) [inline test_package/test_package.cpp]: I don't think we should allow commits which would not compile.

File: .appveyor.yml [hunk 4]
```diff
@@ -1,3 +1,5 @@
+version: "{branch} #{build}
```
Comment (steakhal) [inline .appveyor.yml]: I'm not familiar with the meaning of this line.
What is the benefit of adding this?

Comment (steakhal) [reply .appveyor.yml]: And yea, same here. I wouldn't allow broken stage. I think amending or squashing is the better approach.

File: .travis.yml (hunk 1 above)
Comment (steakhal) [reply .travis.yml]: I would prefer to set everything in the `yml` config file if that is not security-critical.
I think it's better to track setting changes via git as well.

So, in my opinion, this should be here as well.

File: .appveyor.yml (hunk 4 above)
Comment (Neargye) [reply .appveyor.yml]: Agree, I will force push from home. I try commit to pr from phone.

File: test_package/test_package.cpp [hunk 5]
```diff
@@ -14,14 +14,13 @@ int main() {
   static_assert("structvar" == name);
 
   // Nameof.
-  using std::literals::string_view_literals;
-
   std::string_view res1 = NAMEOF(structvar);
   std::string_view res2 = NAMEOF(::structvar);
 
   std::cout << res1 << '\n' << res2 << '\n';
+  using namespace std::literals::string_view_literals;
```
Comment (steakhal) [inline test_package/test_package.cpp]: I don't like `using namespace`. It's much better to say explicitly what you need in your scope.
In this case, I would still stick to this:
```c++
using std::literals::string_view_literals::operator""sv;
```
It isn't much longer but follows this principle.

File: test_package/test_package.cpp [hunk 6]
```diff
@@ -14,14 +14,13 @@ int main() {
...
 
-  bool success = res1 == "structvar" sv && res2 == "structvar" sv;
+  bool success = (res1 == "structvar"sv) && (res2 == "structvar"sv);
```
Comment (steakhal) [inline test_package/test_package.cpp]: There is no need to use parenthesis, but probably you are right and it is more readable in this way. So be it.

File: test_package/test_package.cpp (hunk 5 above)
Comment (Neargye) [reply test_package/test_package.cpp]: Ok, I will revert this when force push soon.

File: .travis.yml (hunk 1 above)
Comment (Neargye) [reply .travis.yml]: Fine, I also like it when I can configure different notifications for different repositories.

File: .appveyor.yml (hunk 4 above)
Comment (Neargye) [reply .appveyor.yml]: By default, only the build number is shown, so it is more convenient to look at which branch is build.

File: .travis.yml (hunk 2 above)
Comment (steakhal) [reply .travis.yml]: According to this [doc](https://docs.travis-ci.com/user/reference/osx/#macos-version) we could use `osx_image: xcode11` as well. That is the most recent on Travis.

Changes pushed containing the [modifications](https://github.com/Neargye/nameof/pull/11/commits/f8cf1c1ea83fd314f4ec3572ae9f02e36d5d55b2).

File: test_package/test_package.cpp (hunk 5 above)
Comment (steakhal) [reply test_package/test_package.cpp]: Feel free to reorder and squash **your** commits.

Comment (steakhal) [reply test_package/test_package.cpp]: Also, it's a good idea to consolidate commit messages.
- capital vs small initial?
- 'dot' at the end of the sentence or blank ending?
- no more `update xyz`, and strictly no commits with the same message in order.

Comment (steakhal) [reply test_package/test_package.cpp]: And you should not `revert` (for now at least).
If you don't feel your git knowledge strong enough to do these changes, I'm here to help you if you ask.

Comment (Neargye) [reply test_package/test_package.cpp]: I agree with you, I have no special wishes.  At work, we have a fairly free format for describing commits, most "fixed #Номер".

**Comment by Neargye** [comment]
Just question, why Linux using docker, and win and macos not 
?

File: test_package/test_package.cpp (hunk 5 above)
Comment (Neargye) [reply test_package/test_package.cpp]: I think the basic approaches for me and you, let use the same:
* small initial
*  blank ending
* no more `update xyz`, and strictly no commits with the same message in order. (update Readme.md - it's ok)

**Comment by Neargye** [comment]
I think ready to merge.

File: test_package/test_package.cpp (hunk 5 above)
Comment (steakhal) [reply test_package/test_package.cpp]: Ok.

**Comment by steakhal** [comment]
Could you tell me why do you have 3 distinct accounts? (usernames: terik23, neargye, Daniil Goncharov)
Don't you want to unify those?

**Comment by steakhal** [comment]
The whole point of using docker - I think - is that you don't have to set up the build toolchain by hand (binutils, compiler, etc)
And you can use a pre-configured one, which is provided by Conan-io (for creating Conan packages).

Why not use docker for Windows and OSX builds?
That's a great question.
First, we have to stick to native environments since we can not virtualize Windows nor MacOSX on Linux.
And somehow I had no toolchain issues on those platforms -- miracle I know.

I don't know much about Docker nor Conan. I'm just a C++ guy/student.

**Comment by Neargye** [comment]
> Could you tell me why do you have 3 distinct accounts? (usernames: terik23, neargye, Daniil Goncharov)
> Don't you want to unify those?

Oh, yes this is a bit confusing. In my community, my two nicks are synonyms, it happened a long time ago (since 2009) and unfortunately not in all accounts I came to a single name. I will correct the settings of the git so that there is one name (unfortunately this will not out quickly, I need to fix about 5 PC).
Daniil Goncharov - my real name.
neargye - nick for most "programming" sites. So if I make commits from home, it applies.
terik23 - nick for work, and most games I play. So if I make commits from work, it applies.

**Comment by Neargye** [comment]
> The whole point of using docker - I think - is that you don't have to set up the build toolchain by hand (binutils, compiler, etc)
> And you can use a pre-configured one, which is provided by Conan-io (for creating Conan packages).
> 
> Why not use docker for Windows and OSX builds?
> That's a great question.
> First, we have to stick to native environments since we can not virtualize Windows nor MacOSX on Linux.
> And somehow I had no toolchain issues on those platforms -- miracle I know.
> 
> I don't know much about Docker nor Conan. I'm just a C++ guy/student.

To be honest, I'm also far from conan. The world of programming is quite extensive and there are quite a few approaches. For example, at work, we still use the Makefile for C projects, although this is the firmware for the controllers, so there are generally quite ancient approaches there. In C++ projects we completely use the сmack, but builds are done on sh scripts.
I am grateful to you for helping me to get acquainted with Conan, I have long wanted to do this.

**Comment by steakhal** [comment]
Btw which do you prefer?
Rebase merging or merge commit?

I think rebase-merging is way better than merge-commits, It's much more readable I think.
If you think the same, I will flatten out the master branch, and force-push.

**Comment by Neargye** [comment]
I like rebase-merge, I think the same :D

**Comment by Neargye** [comment]
For this pr, it's ok merge-commits.

**Comment by steakhal** [comment]
One other tipp from my college that we should use squash-rebase-merge. These changes will be probably small enough to be readable this way.

**Comment by Neargye** [comment]
Earlier, yes, but these changes have already gone public. It seems to me that the ok force-push pr, but not the master branch.

**Comment by steakhal** [comment]
You are probably right.

**Pull request merged**